### PR TITLE
Avoid generating empty reloc section and add ISA name in the .note section

### PR DESCRIPTION
**Pull request by bsaleil**
This patch fixes two minor issues we identified when working on the new part pipeline compilation scheme (https://github.com/GPUOpen-Drivers/llpc/pull/1704/). These two issues are not required for correctness but are nice-to-haves:
* When linking elf files, avoid generating an empty relocation section when there is no relocation.
* Include the ISA name into the `.note` section of the final elf file. This is matching what LLVM is currently doing when not using the part pipeline compilation scheme.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 10d6e0b81f0e354df042acbe25bd7e8a63aaaf6b
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 6d693c7342bb7f18318e80c1da53061dfcd48262</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 2a87cb67e0180322832e722db8a4c53a5fc4e867</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/202608)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67536 (56.6%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29299/67536 (43.4%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 38075/67536 (56.4%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29461/67536 (43.6%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67536 (58.1%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 28301/67536 (41.9%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/elfLinker/ElfLinker.cpp [hunk 1]
```diff
@@ -228,6 +234,7 @@ class ElfLinkerImpl final : public ElfLinker {
   StringMap<unsigned> m_stringMap;                           // Map from string to string table index
   std::string m_notes;                                       // Notes to go in .note section
   bool m_doneInputs = false;                                 // Set when caller has done adding inputs
+  StringRef m_ISAName;                                       // TODO
```
Comment (Flakebi) [inline lgc/elfLinker/ElfLinker.cpp:237]: Comment is TODO? :)

File: lgc/elfLinker/ElfLinker.cpp [hunk 2]
```diff
@@ -692,6 +710,47 @@ void ElfLinkerImpl::mergePalMetadataFromElf(object::ObjectFile &objectFile, bool
...
+  for (const object::SectionRef &section : objectFile.sections()) {
+    object::ELFSectionRef elfSection(section);
+    if (elfSection.getType() == ELF::SHT_NOTE) {
+      Error err = ErrorSuccess();
+      auto &elfFile = cast<object::ELFObjectFile<object::ELF64LE>>(&objectFile)->getELFFile();
+      auto shdr = cantFail(elfFile.getSection(elfSection.getIndex()));
+      for (auto note : elfFile.notes(*shdr, err)) {
+        if (note.getName() == Util::Abi::AmdGpuVendorName && note.getType() == ELF::NT_AMD_HSA_ISA_NAME) {
+          ArrayRef<uint8_t> desc = note.getDesc();
+          m_ISAName = StringRef(reinterpret_cast<const char *>(desc.data()), desc.size());
```
Comment (Flakebi) [inline lgc/elfLinker/ElfLinker.cpp:729]: I think readISAName can return after the first name is found.

File: lgc/elfLinker/ElfLinker.cpp [hunk 3]
```diff
@@ -692,6 +710,47 @@ void ElfLinkerImpl::mergePalMetadataFromElf(object::ObjectFile &objectFile, bool
...
+  noteHeader.n_namesz = noteName.size() + 1;
+  noteHeader.n_descsz = m_ISAName.size();
+  noteHeader.n_type = ELF::NT_AMD_HSA_ISA_NAME;
+  m_notes.append(reinterpret_cast<const char *>(&noteHeader), sizeof(noteHeader));
+  // Write the note name, followed by 1-4 zero bytes to terminate and align.
+  m_notes += noteName;
+  m_notes.append(NoteHeader::Align - (m_notes.size() & NoteHeader::Align - 1), '\0');
+  // Write ISA name, followed by 0-3 zero bytes to align.
+  m_notes += m_ISAName;
+  m_notes.append(-m_notes.size() & NoteHeader::Align - 1, '\0');
```
Comment (Flakebi) [inline lgc/elfLinker/ElfLinker.cpp:751]: That is some interesting bit magic here.

File: lgc/elfLinker/ElfLinker.cpp (hunk 1 above)
Comment (bsaleil) [reply lgc/elfLinker/ElfLinker.cpp]: Thanks :D

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0c61a56590c5c21f30b2081586569f05ab62a8ca
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 6d693c7342bb7f18318e80c1da53061dfcd48262</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 2a87cb67e0180322832e722db8a4c53a5fc4e867</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/202608)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67536 (56.6%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29299/67536 (43.4%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 38075/67536 (56.4%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29461/67536 (43.6%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67536 (58.1%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 28301/67536 (41.9%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 6a01a2f1f4b628ce3e9482c7619e980897f0d67e
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 6d693c7342bb7f18318e80c1da53061dfcd48262</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 2a87cb67e0180322832e722db8a4c53a5fc4e867</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/202608)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67536 (56.6%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29299/67536 (43.4%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 38075/67536 (56.4%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 29461/67536 (43.6%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67536 (58.1%)</li>
<li>Failed: 0/67536 (0.0%)</li>
<li>Not Supported: 28301/67536 (41.9%)</li><li>Warnings: 0/67536 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by kuhar** [comment]
@bsaleil could you silence the typo warning?

> error: `HSA` should be `HAS`

You can do that by updating https://github.com/GPUOpen-Drivers/llpc/blob/dev/.typos.toml

**Comment by amdvlk-admin** [comment]
### Test summary for commit d2e61e2cdf9763fdc986efec47c032a040b092f4
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 6d693c7342bb7f18318e80c1da53061dfcd48262</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 2a87cb67e0180322832e722db8a4c53a5fc4e867</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/202518)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67506 (56.6%)</li>
<li>Failed: 0/67506 (0.0%)</li>
<li>Not Supported: 29269/67506 (43.4%)</li><li>Warnings: 0/67506 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 38075/67506 (56.4%)</li>
<li>Failed: 0/67506 (0.0%)</li>
<li>Not Supported: 29431/67506 (43.6%)</li><li>Warnings: 0/67506 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67506 (58.1%)</li>
<li>Failed: 0/67506 (0.0%)</li>
<li>Not Supported: 28271/67506 (41.9%)</li><li>Warnings: 0/67506 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by bsaleil** [comment]
Added `HSA` to the list of typo corrections.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 51761bc6e419b93751aa66350dd8a29ebbd9ff9e
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 6d693c7342bb7f18318e80c1da53061dfcd48262</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 2a87cb67e0180322832e722db8a4c53a5fc4e867</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/202431)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67477 (56.7%)</li>
<li>Failed: 0/67477 (0.0%)</li>
<li>Not Supported: 29240/67477 (43.3%)</li><li>Warnings: 0/67477 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 38075/67477 (56.4%)</li>
<li>Failed: 0/67477 (0.0%)</li>
<li>Not Supported: 29402/67477 (43.6%)</li><li>Warnings: 0/67477 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67477 (58.1%)</li>
<li>Failed: 0/67477 (0.0%)</li>
<li>Not Supported: 28242/67477 (41.9%)</li><li>Warnings: 0/67477 (0.0%)</li>
</ul></details>


</ul></details>

**Review by piotrAMD — DISMISSED** [review]

File: lgc/elfLinker/ElfLinker.cpp [hunk 4]
```diff
@@ -202,6 +202,12 @@ class ElfLinkerImpl final : public ElfLinker {
   // Read PAL metadata from an ELF file and merge it in to the PAL metadata that we already have
   void mergePalMetadataFromElf(object::ObjectFile &objectFile, bool isGlueCode);
 
+  // Read ISA name string from an ELF file if not already done
+  void readISAName(object::ObjectFile &objectFile);
```
Comment (amdrexu) [inline lgc/elfLinker/ElfLinker.cpp:206]: Please use camelcase `readIsaName` and `writeIsaName`.

**Review by amdrexu — COMMENTED** [review]
Others are fine.

**Comment by bsaleil** [comment]
Updated to use camel case for new functions

**Comment by JaxLinAMD** [comment]
retest this please

**Review by amdrexu — APPROVED** [review]

**Comment by amdvlk-admin** [comment]
### Test summary for commit f25094860ab45dd301e7a3dd4b426b44cfcb3341
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 2483d46fa27c30502e497ea169ee53b142e9fa06</li>
<li>SPVGEN: dev fa39b110650bc72fd7fda54af621e69306036888</li>
<li>XGL: dev 36c5637649089bf65d42c36b431f47c170b57276</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 70268abd53e14b21d8b9960c3fd1e0aefa4dcefb</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/202392)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 38237/67464 (56.7%)</li>
<li>Failed: 0/67464 (0.0%)</li>
<li>Not Supported: 29227/67464 (43.3%)</li><li>Warnings: 0/67464 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 38075/67464 (56.4%)</li>
<li>Failed: 0/67464 (0.0%)</li>
<li>Not Supported: 29389/67464 (43.6%)</li><li>Warnings: 0/67464 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 39235/67464 (58.2%)</li>
<li>Failed: 0/67464 (0.0%)</li>
<li>Not Supported: 28229/67464 (41.8%)</li><li>Warnings: 0/67464 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**